### ISSUE TITLE
Flet APK exit gracefully

### ISSUE THREAD

**Issue by ixenion** [flet-dev/flet#3587]
Im sorry for the issue, because I know it should be rather in discussion, but there is near zero activity there...
How do I make custom "cleanup" logic right before closing (destroying, removing from memory) the android app?

I have one discussion about that, but seems it stuck
(https://github.com/flet-dev/flet/discussions/3555)

The guy from that discussoin advices to use "on_app_lifecycle_state_change"
And this is what I came (but "cleanup" logic does not work):

```python
import flet as ft


def main(page: ft.Page):
    page.title = "Flet counter example"
    page.vertical_alignment = ft.MainAxisAlignment.CENTER

    txt_number = ft.TextField(value="0", text_align=ft.TextAlign.RIGHT, width=100)
    number = page.client_storage.get("number")
    txt_number.value = number if number else "0"


    def minus_click(e):
        txt_number.value = str(int(txt_number.value) - 1)
        page.update()

    def plus_click(e):
        txt_number.value = str(int(txt_number.value) + 1)
        page.update()


    row = ft.Row(
            [
                ft.IconButton(ft.icons.REMOVE, on_click=minus_click),
                txt_number,
                ft.IconButton(ft.icons.ADD, on_click=plus_click),
            ],
            alignment=ft.MainAxisAlignment.CENTER,

        )

    page.add(
            row,
    )


    def handler(e:ft.AppLifecycleStateChangeEvent):
        if e.data == ft.AppLifecycleState.DETACH:
            page.client_storage.set("number", txt_number.value)
            page.window.destroy()

    page.on_app_lifecycle_state_change = handler
    page.window.prevent_close = True

ft.app(main)
```

So It is a simple counter modifyed so that when you exit (close) app it should save last value,
But it doesnt.

The apk compiled successfully, and launches on my phone. All works except that
"ft.AppLifecycleState.DETACH" seems never happen.

What Am I doing wrong?

**Comment by burhansvural** [comment]
> ```python
> AppLifecycleState.DETACH
> ```

https://api.flutter.dev/flutter/dart-ui/AppLifecycleState.html

**Comment by burhansvural** [comment]
```python
import flet as ft


def main(page: ft.Page):
    page.title = "Flet counter example"
    page.vertical_alignment = ft.MainAxisAlignment.CENTER

    txt_number = ft.TextField(value="0", text_align=ft.TextAlign.RIGHT, width=100)
    number = page.client_storage.get("number")
    txt_number.value = number if number else "0"

    def minus_click(e):
        txt_number.value = str(int(txt_number.value) - 1)
        page.update()

    def plus_click(e):
        txt_number.value = str(int(txt_number.value) + 1)
        page.update()

    row = ft.Row(
        [
            ft.IconButton(ft.icons.REMOVE, on_click=minus_click),
            txt_number,
            ft.IconButton(ft.icons.ADD, on_click=plus_click),
        ],
        alignment=ft.MainAxisAlignment.CENTER,

    )

    page.add(
        row,
    )

    def handler(e: ft.AppLifecycleStateChangeEvent):
        print(e.state)
        # ft.AppLifecycleState.DETACH only running on IOS and Android
        if e.state == ft.AppLifecycleState.DETACH:
            page.client_storage.set("number", txt_number.value)
            page.window_destroy()

    page.on_app_lifecycle_state_change = handler
    page.window_prevent_close = True


ft.app(main)

```

**Comment by ixenion** [comment]
> ft.AppLifecycleState.DETACH only running on IOS and Android

So what?
Did you read my question to the end?
I wrote that
"The apk compiled successfully, and launches on my phone.
All works except that
"ft.AppLifecycleState.DETACH" seems never happen."

**Comment by burhansvural** [comment]
I'm sorry I didn't see it.
But the last codes I wrote work without any problems on my Android 12 device.

**Comment by ixenion** [comment]
> @ixenion I'm sorry I didn't see it.
> 
> But the last codes I wrote work without any problems on my Android 12 device. I can only see DETACH working. So the event is triggered, there are deficiencies in the other codes coming from behind, I couldn't deal with them separately.

Did you managed to run it successuffly?
I mean, when you exit (completely close, forse stop, whatever) the app remembers last value
on next startup?

What is your flet and flutte version?
Could you provide link to the apk youv built?

Because I tryed your code, and got same results as mine?

**Comment by burhansvural** [comment]
Yes, it does not save the value at exit. I said this before because there is something missing in its codes.

**Comment by ixenion** [comment]
> Yes, it does not save the value at exit. I said this before because there is something missing in its codes. The issue I was focusing on was whether the DETACH event was triggered or not.

How did you know it worked (DETACH) if the value wasnt saved?

**Comment by ixenion** [comment]
Devs careless attitude is astonishing.
Look https://github.com/flet-dev/flet/issues/3610

![image](https://github.com/flet-dev/flet/assets/48788888/f9a4bf0e-a85e-49a6-a881-c3b5747b1b5a)

He even linked wrong issue. 3566 is not mine and reports about absolutely different thing.

Im afraid with such attitude my "question will be addressed" not very soon

**Comment by ndonkoHenri** [comment]
Sorry for the wrong link.
Except from that, let's keep only this issue opened.

**Comment by ixenion** [comment]
Why it can take so long?
Im not asking to add featire to the lib,
The question is very simple:
Is it possible or not?

**Comment by ixenion** [comment]
> Sorry for the wrong link. Except from that, let's keep only this issue opened.

Why it can take so long?
Im not asking to add featire to the lib,
The question is very simple:
Is it possible or not?

**Comment by ixenion** [comment]
> Yes, it does not save the value at exit. I said this before because there is something missing in its codes. The issue I was focusing on was whether the DETACH event was triggered or not.

Hello?...

**Comment by burhansvural** [comment]
Hi @ixenion

**Comment by ixenion** [comment]
> Hi @ixenion

Dude lol

Iv asked
"How did you know it worked (DETACH) if the value wasnt saved?"

**Comment by ixenion** [comment]
Why no one can read properly?..
I have to repeat questions multiple times,
Its very very very annoying.
Guys c'mon

**Comment by ixenion** [comment]
> Why no one can simply read?.. I have to repeat questions multiple times, Its very very very annoying. Guys c'mon

Thats why Im opening exactly same new issues.

**Comment by ixenion** [comment]
@FeodorFitsner 
Could you check this issue from ths start?
Seems you are in charge of "AppLifecycleState" class:

![Screenshot_20240726-110823_1](https://github.com/user-attachments/assets/aa1203e2-69eb-4f44-b250-e635ba74056a)

![Screenshot_20240726-110841_1](https://github.com/user-attachments/assets/ffa08f35-e89a-4c19-ba25-13174293466c)

![Screenshot_20240726-110708_1](https://github.com/user-attachments/assets/24acfc48-5371-4473-9ffb-40aec902eb71)

**Comment by ixenion** [comment]
It's been half a year since the issue was made...
Any suggestions on how to grasefully exit android app?

**Comment by ixenion** [comment]
Anybody alive?..

**Comment by FeodorFitsner** [comment]
We will look into this issue when its priority increases, if more people are interested in the solution. Adding a lot of comments to the issue won't increase its priority, I'm sorry.

**Comment by ixenion** [comment]
> We will look into this issue when its priority increases, if more people are interested in the solution. Adding a lot of comments to the issue won't increase its priority, I'm sorry.

Hi Feodor!
Do you still think that this is not what people need?

**Comment by ixenion** [comment]
Thanks yall. Really appreciate the participation and engagement.
